lib/backend/etcd.go

```diff
@@ -0,0 +1,229 @@
+		HeaderTimeoutPerRequest: clientTimeout,
+	}
+
+	// Plumb through the username and password if both are configured.
+	if config.EtcdUsername != "" && config.EtcdPassword != "" {
+		cfg.Username = config.EtcdUsername
+		cfg.Password = config.EtcdPassword
+	}
+
+	client, err := etcd.New(cfg)
+	if err != nil {
+		return nil, err
+	}
+	keys := etcd.NewKeysAPI(client)
+
+	return &EtcdClient{etcdClient: client, etcdKeysAPI: keys}, nil
+}
+
+// Create an entry in the datastore.  This errors if the entry already exists.
+func (c *EtcdClient) Create(d DatastoreObject) (*DatastoreObject, error) {
+	return c.set(d, etcdCreateOpts)
+}
+
+// Update an existing entry in the datastore.  This errors if the entry does
+// not exist.
+func (c *EtcdClient) Update(d DatastoreObject) (*DatastoreObject, error) {
+
+	// If the request includes a revision, set it as the etcd previous index.
+	options := etcd.SetOptions{PrevExist: etcd.PrevExist}
+	if d.Revision != nil {
+		options.PrevIndex = d.Revision.(int)
+	}
+
+	return c.set(d, &options)
+}
+
+// Set an existing entry in the datastore.  This ignores whether an entry already
+// exists.
+func (c *EtcdClient) Apply(d DatastoreObject) (*DatastoreObject, error) {
+	return c.set(d, etcdApplyOpts)
+}
+
+// Delete an entry in the datastore.  This errors if the entry does not exists.
+func (c *EtcdClient) Delete(k KeyInterface) error {
+	key, err := k.asEtcdDeleteKey()
+	if err != nil {
+		return err
+	}
+	glog.V(2).Infof("Delete Key: %s\n", key)
+	_, err = c.etcdKeysAPI.Delete(context.Background(), key, &etcdDeleteOpts)
+	return convertEtcdError(err, key)
+}
+
+// Get an entry from the datastore.  This errors if the entry does not exist.
+func (c *EtcdClient) Get(k KeyInterface) (*DatastoreObject, error) {
+	key, err := k.asEtcdKey()
+	if err != nil {
+		return nil, err
+	}
+	glog.V(2).Infof("Get Key: %s\n", key)
+	if results, err := c.etcdKeysAPI.Get(context.Background(), key, &etcdGetOpts); err != nil {
+		return nil, convertEtcdError(err, key)
+	} else if object, err := ParseValue(k, []byte(results.Node.Value)); err != nil {
+		return nil, err
+	} else {
+		return DatastoreObject{Key: k, Object: object, Revision: results.Node.ModifiedIndex}, nil
+	}
+}
+
+// List entries in the datastore.  This may return an empty list of there are
+// no entries matching the request in the ListInterface.
+func (c *EtcdClient) List(l ListInterface) ([]DatastoreObject, error) {
+	// To list entries, we enumerate from the common root based on the supplied
+	// IDs, and then filter the results.
+	key := l.asEtcdKeyRoot()
+	glog.V(2).Infof("List Key: %s\n", key)
+	if results, err := c.etcdKeysAPI.Get(context.Background(), key, &etcdListOpts); err != nil {
+		return nil, err
+	} else {
+		return filterEtcdList(results.Node, l), nil
+	}
+}
+
+// Set an existing entry in the datastore.  This ignores whether an entry already
+// exists.
+func (c *EtcdClient) set(d DatastoreObject, options *etcd.SetOptions) (*DatastoreObject, error) {
```

Comment:
nah - it's just an internal helper method called from the Create/Update/Apply public methods